### PR TITLE
feat: Enable exceptional entries in support rota

### PR DESCRIPTION
**Pull request by kelvin-chappell**
The support rota has a concept of a support duration, which is strictly unnecessary as the structure of the rota can determine the length of a support slot.  It would simplify things if we got rid of it.

There is also a more concrete issue with allowing exceptional entries to be slotted into the rota.  Our current methods for changing support engineers mid-shift, in cases of illness or non-availability, don't work very well.  This is the motivation for making this change.  For context, see issue https://github.com/guardian/janus/issues/1965 and [a recent discussion](https://chat.google.com/room/AAAAfVXX04k/rjuTffeKXiA/rjuTffeKXiA?cls=10) in the 247 general channel.

With this change, if I add, for example, a two-day slot in the middle of a standard support period this will be reflected in the support page and in the support functionality.  So current, next and future slots work as an observer would expect from inspection of the support rota.

The risk is that the rota becomes jumbled and exceptions become the norm.
To mitigate against that we will keep a modified form of the current test that ensures slots begin on Tuesdays but allow an 'imminent' period in which slots can be more flexible.

Tested locally and Janus data ingestion works with and without a support duration.
Support page and functionality is unchanged.

### CODE REVIEW

File: test/logic/UserAccessTest.scala [hunk 1]
```diff
@@ -2,12 +2,12 @@ package logic
 
 import com.gu.googleauth.UserIdentity
 import com.gu.janus.model.{ACL, SupportACL}
-import fixtures.Fixtures._
+import fixtures.Fixtures.*
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply test/logic/UserAccessTest.scala:5]: In Scala 3, wildcard imports are expressed with '*' rather than '_'.
https://docs.scala-lang.org/scala3/reference/changed-features/imports.html

File: app/logic/UserAccess.scala [hunk 2]
```diff
@@ -86,15 +95,22 @@ object UserAccess {
...
-          (startTime, if (user1 == user) user2 else user1)
-      }
-      .filter { case (start, _) =>
-        start.isAfter(date.plus(supportACL.supportPeriod))
+    val nextSlotStartTime =
+      nextSupportUsers(date, supportACL).map((startTime, _) => startTime)
+    nextSlotStartTime
+      .map { nextSlot =>
+        supportACL.rota.toList
+          .sortBy((startTime, _) => startTime)
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply app/logic/UserAccess.scala:103]: It doesn't seem like a big problem given the size of the list.
Perhaps an optimisation for later.

File: app/logic/UserAccess.scala [hunk 3]
```diff
@@ -50,30 +50,39 @@ object UserAccess {
...
 
   def nextSupportUsers(
       date: Instant,
       supportACL: SupportACL
   ): Option[(Instant, (Option[String], Option[String]))] = {
-    activeSupportUsers(date.plus(Duration.ofDays(7)), supportACL)
+    supportACL.rota.toList
+      .sortBy((startTime, _) => startTime)
+      // First slot start-time after or at given instant
+      .find((startTime, _) => startTime.isAfter(date))
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply app/logic/UserAccess.scala:79]: It doesn't seem like a big problem given the size of the list.
Perhaps an optimisation for later.

**Review by emdash-ie — APPROVED** [review]
I’m not very familiar with the code, but it looks good to me! Thanks for doing this – I had it on my list but never got to it :)

File: test/logic/UserAccessTest.scala [hunk 4]
```diff
@@ -252,6 +308,47 @@ class UserAccessTest
           val (startTime, _) = activeSupportUsers(currentTime, acl).value
           startTime shouldEqual rotaStartTime
         }
+
+        "should always be defined if the support rota has a slot before the given time" in {
```
Comment (adamnfish-gu) [inline test/logic/UserAccessTest.scala]: Can we focus this a bit more on the meaning, rather than the actual data? The test's source code already tells us that we're looking for a "defined" value, no need to duplicate that here. I think the intent is something like "rotas last forever, so however far back in history it is the most recent support rota entry is in effect". In more terse English though, if you agree?

File: test/logic/UserAccessTest.scala [hunk 5]
```diff
@@ -252,6 +308,47 @@ class UserAccessTest
...
+        }
+
+        "should never be defined if the support rota has no slots before the given time" in {
+          forAllCases(rotaAndTimeBeforeRotaTimespan) {
+            case (acl: SupportACL, time: Instant) =>
+              activeSupportUsers(time, acl).isEmpty shouldBe true
+          }
+        }
+
+        "should always be in a support slot that became active before the given time" in {
```
Comment (adamnfish-gu) [inline test/logic/UserAccessTest.scala]: I'm struggling to see how this significantly differs from the first test? I'm also a bit confused by this one from the wording though. We wouldn't be "in" a support slot if it's in the past and another one has kicked in since then?

File: test/logic/UserAccessTest.scala [hunk 6]
```diff
@@ -219,11 +228,58 @@ class UserAccessTest
...
 
+      val engineers = List("eng1", "eng2", "eng3", "eng4", "eng5")
+
+      val genEngineerPair = for {
+        e1 <- Gen.oneOf(engineers)
+        e2 <- Gen.oneOf(engineers.filter(_ != e1))
+      } yield (e1, e2)
+      val genSupportSlotCount = Gen.choose(min = 3, max = 10)
+
+      def genInstant(min: Instant, max: Instant) = Gen
```
Comment (adamnfish-gu) [inline test/logic/UserAccessTest.scala]: I'd just want to make sure these constraints are obeyed in a failing test, when scalacheck starts its "shrinking" process. Seems ok from my tinkering, but as ever, make sure you've seen these tests fail and give helpful feedback!

**Review by adamnfish-gu — COMMENTED** [review]
You might want to update the janusData.conf files in the test resources directory of the configTools project as well? These still contain the support period, now no longer needed.
e.g. https://github.com/guardian/janus-app/blob/92b7b746c3b8ec41ce32c39d17ef97060bc11461/configTools/src/test/resources/example-without-permissions-repo.conf#L32

A few comments, we can chat in person as well.

Overall the approach looks fine to me!

File: test/logic/UserAccessTest.scala [hunk 7]
```diff
@@ -273,20 +370,34 @@ class UserAccessTest
...
-        }
-
         "returns the date the rota started at" in {
           val acl = testActiveSupportAcl("user.1", "user.2")
           val (startTime, _) =
             nextSupportUsers(currentTimeForNextRota, acl).value
           startTime shouldEqual rotaStartTime
         }
+
+        "should always be defined at any time within the timespan of the support rota" in {
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply test/logic/UserAccessTest.scala]: Improved most confusing test titles in f5a83c8754d3933a22ac5a5e268c3a26b5d58c9f

File: test/logic/UserAccessTest.scala [hunk 8]
```diff
@@ -338,6 +448,41 @@ class UserAccessTest
           val slots = futureRotaSlotsForUser(currentTime, supportAcl, "userA")
           slots.isEmpty shouldBe true
         }
+
+        "should always be after the given time for all engineers" in {
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply test/logic/UserAccessTest.scala]: Improved in f5a83c8754d3933a22ac5a5e268c3a26b5d58c9f
